subxt/src/config/signed_extensions.rs

```diff
@@ -10,19 +10,45 @@
 use super::extrinsic_params::{ExtrinsicParams, ExtrinsicParamsEncoder, ExtrinsicParamsError};
 use crate::utils::Era;
 use crate::{client::OfflineClientT, Config};
-use codec::{Compact, Encode};
+use codec::{Compact, Decode, Encode};
 use core::fmt::Debug;
+use scale_decode::DecodeAsType;
 use std::collections::HashMap;
 
 /// A single [`SignedExtension`] has a unique name, but is otherwise the
 /// same as [`ExtrinsicParams`] in describing how to encode the extra and
 /// additional data.
-pub trait SignedExtension<T: Config>: ExtrinsicParams<T> {
+pub trait SignedExtension<T: Config>: ExtrinsicParams<T> + SignedExtensionDecoder {
     /// The name of the signed extension. This is used to associate it
     /// with the signed extensions that the node is making use of.
     const NAME: &'static str;
 }
 
+/// Specifies the Extra and Additional data types of the signed extension.
+/// `ExtrinsicParamsEncoder` is implemented for all `SignedExtensionDecoder`s.
+pub trait SignedExtensionDecoder {
+    /// Signed Extra parameter of a Signed Extension.
+    /// Included in a signed extension behind the signature.
+    type Extra: Encode + DecodeAsType;
+    /// Additional Signed parameter of a Signed Extension.
+    /// Used as part of the payload for signing an extrinsic.
+    type Additional: Encode;
+    /// Retrieves the signed extra parameter.
+    fn extra(&self) -> &Self::Extra;
+    /// Retrieves the additional signed parameter.
+    fn additional(&self) -> &Self::Additional;
+}
+
+impl<S: SignedExtensionDecoder + 'static> ExtrinsicParamsEncoder for S {
+    fn encode_extra_to(&self, v: &mut Vec<u8>) {
+        self.extra().encode_to(v);
+    }
+
+    fn encode_additional_to(&self, v: &mut Vec<u8>) {
+        self.additional().encode_to(v);
+    }
+}
+
```

Comment:
I'll have to think about this a bit!

On the one hand, I do like the symmetry between `extra` and `additional`, and the `find` API is lovely!

On the other hand, I have some reservations still:
- `ExtrinsicParamsEncoder` is in a weird place now. The blanket impl means that nobody can create their own impl of it now I think (so if I wanted to impl `ExtrinsicParams` myself and ignore the signed extension stuff, I couldn't). It also just doesn't fit nicely into the picture now and is relegated to being a hack for object safety.
- Previously, `ExtrinsicParamsEncoder` allowed one to be very flexible about how they encode extra+additional properties, because how the encoding happens is hidden behind a method. They could use plain old `codec::Encode`, or they could write a signed extension which stores the metadata and then uses that to more intelligently encode properties based on metadata. 
  - There are definite uses for this; `ChargeAssetTxPayment` for instance can be generic over asset location, and metadata would allow somebody to be smart and write a version of it that worked in a few places. S
  - Since a signed extension can save the metadata, maybe this is ok though; they can just have a custom `Encode` impl which internally uses metadata to do something smarter. Hmmm.
- Currently we have `type Extra = Encode + DecodeAsType` which is a weird asymmetry, though I can't see a way around that offhand (Encode doesn't need the metadata perhaps, but decoding does).
- The requirement to point to a separate type for the `Additional` thing too exists only the symmetry, but does it actually benefit us or just make life a little harder for no reason?

I think the main thing in my head to resolve then is the weird place that `ExtrinsicParamsEncoder` is in, and whether it's still possible to allow a user to implement `ExtrinsicParams` themselves without having to go via signed extensions (I'm not sure how much this would be useful, but it's a nice escape hatch to allow users to add their very own abstractions instead if our SignedExtension stuff isn't working for them).

I do still wonder whether it is better to just add a `type DecodedExtra: DecodeAsType` bound to `SignedExtension` and rely on that; this doesn't have the symmetry (but maybe that's misleading anyway), but it does mean that `ExtrinsicParams` can still be implemented normally, avoids needing _another_ trait here (there are already 3 :D) and leaves `ExtrinsicParamsEncoder` with a clear purpose other than to be an object safety hack. I think it'd be good to see what that option looks like as well, perhaps in a separate PR, for comparison :)